File: frontend/apps/ios/Pantopus/Core/Networking/Models/IdentityCenter/IdentityCenterDTOs.swift
//
//  IdentityCenterDTOs.swift
//  Pantopus
//
//  Decoder shapes for `/api/identity-center/*`. Backend column names
//  stay (persona / bridges); the UI renames at the view-model
//  boundary. See docs/identity-firewall-ui-ux-redesign-2026-05-06.md.
//

import Foundation

/// Envelope from `GET /api/identity-center`.
public struct IdentityCenterResponse: Decodable, Sendable {
    public let privateAccount: PrivateAccountDTO?
    public let localProfile: LocalProfileDTO?
    public let audienceProfile: AudienceProfileDTO?
    public let bridges: BridgesDTO?
    public let homes: [HomeIdentityDTO]?
    public let businessProfiles: [BusinessIdentityDTO]?
    public let personaCount: Int?
    public let blockCounts: BlockCountsDTO?

    public init(
        privateAccount: PrivateAccountDTO?,
        localProfile: LocalProfileDTO?,
        audienceProfile: AudienceProfileDTO?,
        bridges: BridgesDTO?,
        homes: [HomeIdentityDTO]?,
        businessProfiles: [BusinessIdentityDTO]?,
        personaCount: Int?,
        blockCounts: BlockCountsDTO?
    ) {
        self.privateAccount = privateAccount
        self.localProfile = localProfile
        self.audienceProfile = audienceProfile
        self.bridges = bridges
        self.homes = homes
        self.businessProfiles = businessProfiles
        self.personaCount = personaCount
        self.blockCounts = blockCounts
    }

    enum CodingKeys: String, CodingKey {
        case privateAccount = "private_account"
        case localProfile = "local_profile"
        case audienceProfile = "audience_profile"
        case bridges, homes
        case businessProfiles = "business_profiles"
        case personaCount = "persona_count"
        case blockCounts = "block_counts"
    }
}

/// The signed-in user's email-level identity. Surfaces as "Personal"
/// in the design.
public struct PrivateAccountDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let email: String?
    public let name: String?
    public let verified: Bool?
    public let profilePictureUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, email, name, verified
        case profilePictureUrl = "profile_picture_url"
    }
}

/// "Local Profile" card data.
public struct LocalProfileDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let handle: String?
    public let displayName: String?
    public let avatarUrl: String?
    public let postCount: Int?
    public let connectionCount: Int?
    public let verified: Bool?
    public let locality: String?

    enum CodingKeys: String, CodingKey {
        case id, handle
        case displayName = "display_name"
        case avatarUrl = "avatar_url"
        case postCount = "post_count"
        case connectionCount = "connection_count"
        case verified, locality
    }
}

/// "Public profile" card (backend calls it audience profile / persona
/// — renamed at the UI boundary).
public struct AudienceProfileDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let handle: String?
    public let displayName: String?
    public let avatarUrl: String?
    public let followerCount: Int?
    public let postCadence: String?
    public let status: String?

    enum CodingKeys: String, CodingKey {
        case id, handle
        case displayName = "display_name"
        case avatarUrl = "avatar_url"
        case followerCount = "follower_count"
        case postCadence = "post_cadence"
        case status
    }
}

/// Per-identity bridge toggles — "Profile links" in the UI.
public struct BridgesDTO: Decodable, Sendable, Hashable {
    public let showPersonaOnLocal: Bool?
    public let showLocalOnPersona: Bool?

    enum CodingKeys: String, CodingKey {
        case showPersonaOnLocal = "show_persona_on_local"
        case showLocalOnPersona = "show_local_on_persona"
    }
}

/// One row in the Homes list.
public struct HomeIdentityDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let name: String?
    public let city: String?
    public let state: String?
    public let primaryPhotoUrl: String?
    public let role: String?

    enum CodingKeys: String, CodingKey {
        case id, name, city, state
        case primaryPhotoUrl = "primary_photo_url"
        case role
    }
}

/// One row in the Business profiles list.
public struct BusinessIdentityDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let displayName: String?
    public let role: String?
    public let isActive: Bool?

    enum CodingKeys: String, CodingKey {
        case id
        case displayName = "display_name"
        case role
        case isActive = "is_active"
    }
}

public struct BlockCountsDTO: Decodable, Sendable, Hashable {
    public let personal: Int?
    public let audience: Int?
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/identity/IdentityCenterDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.identity

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/** Envelope from `GET /api/identity-center`. */
@JsonClass(generateAdapter = true)
data class IdentityCenterResponse(
    val privateAccount: PrivateAccountDto? = null,
    val localProfile: LocalProfileDto? = null,
    val audienceProfile: AudienceProfileDto? = null,
    val bridges: BridgesDto? = null,
    val homes: List<HomeIdentityDto>? = null,
    val businessProfiles: List<BusinessIdentityDto>? = null,
    val personaCount: Int? = null,
    val blockCounts: BlockCountsDto? = null,
)

@JsonClass(generateAdapter = true)
data class PrivateAccountDto(
    val id: String,
    val email: String? = null,
    val name: String? = null,
    val verified: Boolean? = null,
    @Json(name = "profile_picture_url") val profilePictureUrl: String? = null,
)

@JsonClass(generateAdapter = true)
data class LocalProfileDto(
    val id: String,
    val handle: String? = null,
    @Json(name = "display_name") val displayName: String? = null,
    @Json(name = "avatar_url") val avatarUrl: String? = null,
    @Json(name = "post_count") val postCount: Int? = null,
    @Json(name = "connection_count") val connectionCount: Int? = null,
    val verified: Boolean? = null,
    val locality: String? = null,
)

@JsonClass(generateAdapter = true)
data class AudienceProfileDto(
    val id: String,
    val handle: String? = null,
    @Json(name = "display_name") val displayName: String? = null,
    @Json(name = "avatar_url") val avatarUrl: String? = null,
    @Json(name = "follower_count") val followerCount: Int? = null,
    @Json(name = "post_cadence") val postCadence: String? = null,
    val status: String? = null,
)

@JsonClass(generateAdapter = true)
data class BridgesDto(
    @Json(name = "show_persona_on_local") val showPersonaOnLocal: Boolean? = null,
    @Json(name = "show_local_on_persona") val showLocalOnPersona: Boolean? = null,
)

@JsonClass(generateAdapter = true)
data class HomeIdentityDto(
    val id: String,
    val name: String? = null,
    val city: String? = null,
    val state: String? = null,
    @Json(name = "primary_photo_url") val primaryPhotoUrl: String? = null,
    val role: String? = null,
)

@JsonClass(generateAdapter = true)
data class BusinessIdentityDto(
    val id: String,
    @Json(name = "display_name") val displayName: String? = null,
    val role: String? = null,
    @Json(name = "is_active") val isActive: Boolean? = null,
)

@JsonClass(generateAdapter = true)
data class BlockCountsDto(
    val personal: Int? = null,
    val audience: Int? = null,
)

/** Body for `PATCH /api/identity-center/bridges/:personaId`. */
@JsonClass(generateAdapter = true)
data class UpdateBridgesBody(
    @Json(name = "show_persona_on_local") val showPersonaOnLocal: Boolean? = null,
    @Json(name = "show_local_on_persona") val showLocalOnPersona: Boolean? = null,
)

/** Acknowledgement envelope from the PATCH. */
@JsonClass(generateAdapter = true)
data class BridgesEchoResponse(
    val bridges: BridgesDto? = null,
)
